_posts/2020-09-30-hows-the-performance-of-ruby-3.0.0-preview1.md

```diff
@@ -0,0 +1,80 @@
+
+And now, using version 3.0 with the serial numbers filed off, off we go!
+
+Nearly. In fact, it looks like my (badly-patched) version has an incompatibility between (un-reverted) Ruby 3.0 changes and a monkeypatch claiming to be for ActionModel::Serializer 0.8/0.9 (!). So after removing a (very small) bit of functionality in favour of returning a static string, and turning down the number of load threads a bit to avoid an occasional segfault... **Now** off we go.
+
+## Software for Testing
+
+My existing test stack seemed like the way to go: I've built this very specifically to time Ruby 3.0 against Ruby 2.0, for many years. Now that Ruby 3 exists, I can finally use it for that! It's about time!
+
+[Rails Ruby Bench](https://github.com/noahgibbs/rails_ruby_bench) runs a copy of [Discourse](https://github.com/discourse/discourse), a common and popular Rails app to host internet forums. It's one of the biggest available "real" open-source Rails apps, making it a fine choice for "real world" benchmarking. RRB runs a set of simulated pseudorandom user requests against the running Rails app, and times how long they all take to finish. So it's a throughput test. [You can run it yourself if you like](https://engineering.appfolio.com/appfolio-engineering/2019/11/28/how-do-i-use-rails-ruby-bench), though it's a bit complex and finicky. You may be gathering that from the "Pitfalls" section above. The dark side of using real-world software is hitting real-world complexity and bugs.
+
+RRB has run against AWS m4.2xlarge instances for nearly its whole existence. That's what I'm doing again. For now: m4.2xlarge instances, 10 processes, 6 threads/process, the same as I've been [using for over 3 years](https://rubykaigi.org/2017/presentations/codefolio.html) for this purpose. Once Ruby 3 comes out, it'll be time to look at upgrading RRB. After that, it will have fulfilled its purpose: to measure the total speedup from 2.0 to 3.0 of a typical real-world Ruby on Rails application.
+
+And I can finally stop supporting some hideous old hacks to make that possible.
+
+These results are running minimum 30 batches per Ruby version of 10,000 requests each. That's enough to get pretty solid results, but not to detect tiny differences.
+
+(Due to randomization and a restart after fixing a rare bug, we got 38 batches for Ruby 2.7 and 36 batches for Ruby 3.0.0-preview1.)
+
+## Results
+
+How are results? Not dramatic, I'm afraid.
+
+<table>
+    <thead>
+        <tr><td></td><th>Ruby 2.7</th><th>Ruby 3.0.0-preview1</th><th>Speedup/Slowdown</th></tr>
+    </thead>
+    <tbody>
+        <tr>
+            <th>Median Throughput</th><td>165.7</td><td>160.7</td><td style="color:red">-3.0%</td>
+            <th>Fastest Run of 30</th><td>168.7</td><td>164.0</td><td style="color:red">-2.8%</td>
+            <th>Slowest Run of 30</th><td>163.0</td><td>158.2</td><td style="color:red">-2.9%</td>
+        </tr>
+    </tbody>
+</table>
```

Comment:
@noahgibbs Maybe we can break this table in two. It looks a little weird right now:

![Screen Shot 2020-09-30 at 12 17 28 PM](https://user-images.githubusercontent.com/7442887/94705012-ef118a00-0316-11eb-9719-b352c4e09475.png)

Reply:
Aaaaand that's a cut-and-paste error that I couldn't see because I was having trouble building :-/ I put it into another Jekyll project and make sure the table looked right. Not perfect, but definitely better now.